smdebug/tensorflow/keras.py

```diff
@@ -620,3 +621,75 @@ def new_add_weight(opt, *args, **kwargs):
         # Optimizer is being saved to support additional features in the future.
         self.optimizer = optimizer
         return optimizer
+
+    def forward_pre_hook(self, tape):
+        self.tape = tape
+        self.set_mode(ModeKeys.TRAIN)
+        if self._is_not_supported():
+            return
+        self.worker = self._get_worker_name()
+        if self.writer is not None or len(self.writer_map):
+            self._close_writers()
+        if not self.prepared_collections:
+            # at this point we need all collections to be ready
+            # this may not be the case at creation of hook
+            # as user's code after hook might add collections
+            self._prepare_collections()
+            self.prepared_collections = True
+
+        self._increment_step()
+
+        if self._get_collections_to_save_for_step():
+            self._initialize_writers()
+
+        if self._exported_collections is False:
+            # in keras, these collections change when mode changes
+            # but rest of the project isn't yet capable of handling this
+            # this means that collections like outputs, or other collections with intermediate tensors
+            # will only have tensor names from first mode
+
+            # this means sometimes collections will be exported after 1 step
+            self.export_collections()
+            self._exported_collections = True
+
+    def update_step(self, grads, vars):
+        if self._get_collections_to_save_for_step():
+            if grads is not None and vars is not None:
+                for idx, (g, v) in enumerate(zip(grads, vars)):
+                    layer = v.name.split(":")[0]
+                    self._save_for_tensor(
+                        tensor_name="gradients/" + layer + "Grad",
+                        tensor_value=g,
+                        check_before_write=True,
+                    )
+                    self._save_for_tensor(
+                        tensor_name="weights/" + v.name,
+                        tensor_value=v.value(),
+                        check_before_write=True,
+                    )
+
+    def record_tensor_value(self, tensor_name, tensor_value, tape):
+        self.tape = tape
+
+        def add_metric_to_coll(self, tensor_name, tensor_value):
+            if tensor_name not in self.tensor_to_collections:
+                matched_colls = set()
+                if tensor_name in ["loss", "val_loss"]:
+                    coll_name = CollectionKeys.LOSSES
+                else:
+                    coll_name = CollectionKeys.METRICS
```

Comment:
Seems like this logic should be covered by the CollectionKey regexes. Can you elaborate why we're hardcoding this instead of referring to existing collections methods?

Reply:
there's a regex for Loss but i dont think there's a regex for metrics

Reply:
Might be worth adding one then. There's significant codebase complexity that comes from passing around collections, seeing if collections are prepared, and seeing if tensors match the collection - duplicated all over the place. The more we can refactor and simplify, the better :)

Reply:
let me try, this might require enforcing that the metric name includes a "metric" word. 

reused the existing add_metric() in Keras hook for now.